devlib/utils/ssh.py

```diff
@@ -170,51 +305,422 @@ def name(self):
+        }
+
+        def redirect_thread_f(stdout_in, stderr_in, out_streams, select_timeout):
+            def callback(out_streams, name, chunk):
+                try:
+                    r, w = out_streams[name]
+                except KeyError:
+                    return out_streams
+
+                try:
+                    w.write(chunk)
+                # Write failed
+                except ValueError:
+                    # Since that stream is now closed, stop trying to write to it
+                    del out_streams[name]
+                    # If that was the last open stream, we raise an
+                    # exception so the thread can terminate.
+                    if not out_streams:
+                        raise
+
+                return out_streams
+
+            try:
+                _read_paramiko_streams(stdout_in, stderr_in, select_timeout, callback, copy.copy(out_streams))
+            # The streams closed while we were writing to it, the job is done here
+            except ValueError:
+                pass
+
+            # Make sure the writing end are closed proper since we are not
+            # going to write anything anymore
+            for r, w in out_streams.values():
+                if r is not w and w is not None:
+                    w.close()
+
+        # If there is anything we need to redirect to, spawn a thread taking
+        # care of that
+        select_timeout = 1
+        thread_out_streams = {
+            name: (r, w)
+            for name, (r, w) in out_streams.items()
+            if w is not None
+        }
+        redirect_thread = threading.Thread(
+            target=redirect_thread_f,
+            args=(stdout_in, stderr_in, thread_out_streams, select_timeout),
+            # The thread will die when the main thread dies
+            daemon=True,
+        )
+        redirect_thread.start()
+
+        return ParamikoBackgroundCommand(
+            conn=self,
+            as_root=as_root,
+            chan=channel,
+            pid=pid,
+            stdin=stdin,
+            # We give the reading end to the consumer of the data
+            stdout=out_streams['stdout'][0],
+            stderr=out_streams['stderr'][0],
+            redirect_thread=redirect_thread,
+        )
+
+    def _close(self):
+        logger.debug('Logging out {}@{}'.format(self.username, self.host))
+        with _handle_paramiko_exceptions():
+            bg_cmds = set(self._current_bg_cmds)
+            for bg_cmd in bg_cmds:
+                bg_cmd.close()
+            self.client.close()
+
+    def _execute_with_sudo(self, command, as_root, log, timeout, executor):
+        # As we're already root, there is no need to use sudo.
+        log_debug = logger.debug if log else lambda msg: None
+        use_sudo = as_root and not self.connected_as_root
+
+        if use_sudo:
+            if not self.password:
+                raise TargetStableError('Attempt to use sudo but no password was specified')
```

Comment:
No password is prompted since that's not an interactive shell. This means we use `sudo -S` to read the password from stdin. We could make check for `None` though:
```
if self.password is None:
     raise ....
```
Which will allow passing an empty string as password.

Reply:
That seems like it might be a better option. Another thought out of curiosity, do you know what state occurs if we just don't provide the password after invoking sudo (if no password is provided) and is that state easy to detect?

Reply:
Good question, now that I think about it, passwordless sudo might just execute the command without even looking at stdin, in which case we must not supply one (otherwise the called command will receive a newline it never asked for). Do you have  a passwordless system where you can check the behavior of `sudo -S` ?

Reply:
Just checked on my chrombook and that is the case, it immediately just executes the command.

Reply:
So now the situation is:
* password=None means we don't provide a password. If as_root is used, it will raise an exception
* password="" means sudo is configured to run passwordless. If sudo actually needs a password, it will stall waiting on stdin.

The only way to avoid the stall in case of misconfiguration would be to send "\n" in all cases. Unfortunately, that means sending a newline to a command that did not ask for it when everything is well configured. All that can be avoided by a check at connection time to see if sudo is passwordless or not. Should we go that way ?

Reply:
If we can check this once at connection time that seems like a good plan.
Although if we do this check, does it still make sense to make the distinction between providing an empty password vs no password?

Reply:
I think it's a good idea to allow users to disable the use of sudo if they want to. That could be used to prevent devlib from poking into important things without knowing about it.

Reply:
Ah ok I hadn't thought about using that as a feature.

In which case just to clarify the above:
> password=None means we don't provide a password. If as_root is used, it will raise an exception

Will this still allow for the usecase where we're connected as root (using a private key, no password) and calling `execute` with `as_root=True` since then `sudo` wouldn't be required?

Reply:
good point, `password` is not a good indicator of what the user wants. I guess we should just check if it's passwordless sudo or not and just check for `if self.password:` as you suggested.